### ISSUE TITLE
asm.js/wasm support

### ISSUE THREAD

**Issue by icefoxen** [ggez/ggez#71]
Should happen eventually!

**Comment by icefoxen** [comment]
See https://github.com/tomaka/cpal/issues/92

**Comment by icefoxen** [comment]
Blockers:

* Resources and files
* Audio
* SDL GL context attributes: https://github.com/kripken/emscripten/issues/4425
* gfx-rs support: When I try to run a hacked-up test program I get tons of the following errors: `bad name in getProcAddress: glWhateverThing`

**Comment by kvark** [comment]
@icefoxen gfx-0.17 runs on WebGL, but a few tweaks are required in the environment. See (from https://github.com/gfx-rs/gfx/pull/1681):
> In Emscripten configuration, set USE_WEBGL2=1. I did it by just modifying "emscripten/src/settings.js" directly (from where it's installed).

**Comment by beatgammit** [comment]
Since `ggez` is using `winit` now, the SDL bit is irrelevant, right? Aside from that, is the list above accurate?

**Comment by icefoxen** [comment]
Yeah, but `winit` support for `wasm32-unknown` is still experimental last I checked.  And the stable version of ggez still uses SDL2, so.

**Comment by icefoxen** [comment]
Currently, here are the crates that need to support wasm:

 * `slice_deque`: https://github.com/gnzlbg/slice_deque/issues/60 (though `minimp3`, which is easy to get rid of by disabling the `mp3` feature in `rayon)
 * `bzip2`, used by `zip`, again easy to feature out
 * `shared_library`: used by `glutin` and by `osmesa-sys` which is used by `glutin`
 * `winit`: https://github.com/tomaka/winit/issues/395

**Comment by kvark** [comment]
Worth noting this project that has a dedicated WebGL backend on GGEZ: http://github.com/not-fl3/good-web-game

**Comment by ozkriff** [comment]
Note that good-web-game is not really GGEZ's backend, but a separate web-targeted engine with a similar API that @not-fl3 uses for [his prototypes](https://twitter.com/notfl3/status/1079499336243965952). Zemeroth uses good-web-game for its web version as a quick-n-dirty immediate solution until a proper WASM support arrives to GGEZ (there're no plans of making good-web-game some kind of official GGEZ backend). The currently implemented subset of GGEZ API is quite limited and while it may be used for something else that Zemeroth, it will probably require a lot of work to do (contributions are welcome ;) ).

**Comment by Leinnan** [comment]
@icefoxen  According to wiki winit is the main blocker for WASM support. The issue You did mention looks as resolved: https://github.com/rust-windowing/winit/issues/395

**Comment by icefoxen** [comment]
Yep, `winit` 0.20 ought to have wasm support!  I'm looking forward to giving it a go and seeing what happens.

**Comment by icefoxen** [comment]
`winit` actually works decently now, kinda, but `gfx-hal` turns out to be something that doesn't add much value, at least not yet.

**Comment by JohnnyUrosevic** [comment]
So are we just waiting on other crates to support WASM to make this functionality happen?

**Comment by icefoxen** [comment]
@JohnnyUrosevic Nah we're waiting for me (or someone) to finish the `ggraphics` branch and rewrite the entire graphics stack.  XD

**Comment by mystal** [comment]
@icefoxen Is there an issue with details about a desired path forward for ggez's graphics stack? I found this that mentions switching to glow, but would love more context: https://github.com/ggez/ggez/issues/418#issuecomment-562997102

Reason being that I'd love to contribute :)

**Comment by icefoxen** [comment]
It's been through a few awkward twists and turns, is part of the issue.  It started out as "of course we're going to use gfx-hal and rendy".  Then after like three months of slowly and awkwardly making that Good but then realizing it still won't work well on web for the forseeable future I ditched it and switched to just using OpenGL with `glow`, which is the `ggraphics` branch.  The readme in https://github.com/ggez/ggez/tree/ggraphics/ggraphics has the general design overview.  I've gotten thoroughly distracted with other things the last few months, annoyingly, but have been thinking that maybe it's time to start taking it seriously again.  So, please ask if you have any other questions.

**Comment by JohnnyUrosevic** [comment]
is there any low hanging fruit ways to help? pretty new to rust as a whole but rust + wasm honestly seems like the future for many things, especially games

**Comment by mystal** [comment]
I guess my first question is why not use [`miniquad`](https://github.com/not-fl3/miniquad)? Or rather, how is `ggraphics` different from `miniquad`?

I've used `good-web-game` (pre-miniquad) to run my ggez game on WASM and found that it worked decently well. I figured since good-web-game now uses miniquad it must be at least somewhat decent :)

Happy to do some digging on that front if the answer is "I haven't thought about it much".

**Comment by not-fl3** [comment]
We had a discussion about possibilities to use some of the work from `good-web-game` a while ago and @icefoxen had a bit different view on windowing/graphics stack.

However, I am always ready to open a giant pull request to `ggez` and nuke `good-web-game` :)

**Comment by FeldrinH** [comment]
Now that ggez has new maintsiners, is there any update on when and how wasm support might happen?

**Comment by PSteinhaus** [comment]
We're currently discussing what to replace pre-II-gfx with.
The `ggraphics` branch (using `glow`) is one option. Other options are `miniquad` and maybe `wgpu`.

Once we replaced gfx we can try to get wasm to work with whatever we replaced it with. But before we dive into that I guess we really have to weigh the pros and cons for a bit and see what we're most comfortable with.

So, there's no real update on how, but as to "when" I think we can answer "next (probably)".

**Comment by not-fl3** [comment]
Hi! `miniquad` maintainer here :) 

https://github.com/not-fl3/good-web-game - API-compatible `ggez` implementation built with `miniquad` still works, however, I shifted my focus towards `macroquad` and do not actively maintain it.

Some people still use it tho, as an example - https://github.com/eduardonunesp/rust-sokoban-wasm ggez sokoban game runned in the browser with `good-web-game`. 

If `miniquad` will be chosen as a rendering backend - I would be happy to get back to `gwg` and help with doing a PR from `gwg` to `ggez` :)

**Comment by nobbele** [comment]
We are moving the discussion to #962 by the way.

**Comment by PSteinhaus** [comment]
While Wasm is still not supported by ggez directly I want to mention that [good-web-game](https://github.com/ggez/good-web-game) has now been expanded and updated to encompass all of the core functionality of ggez 0.6 that is portable to the web.

So while classic ggez still lacks Wasm support good-web-game now enables anyone to port their ggez-0.6 games to Android, iOS or Wasm without too much hassle.
